util/oci/client.go

```diff
@@ -0,0 +1,544 @@
+	if disableManifestMaxExtractedSize {
+		maxSize = math.MaxInt64
+	}
+
+	manifestsDir, err := extractContentToManifestsDir(ctx, cachedPath, digest, maxSize, c.allowedMediaTypes)
+	if err != nil {
+		return manifestsDir, nil, err
+	}
+
+	return manifestsDir, argoio.NewCloser(func() error {
+		return os.RemoveAll(manifestsDir)
+	}), nil
+}
+
+func (c *nativeOCIClient) getCachedPath(version, project string) (string, error) {
+	keyData, err := json.Marshal(map[string]string{"url": c.repoURL, "project": project, "version": version})
+	if err != nil {
+		return "", err
+	}
+	return c.repoCachePaths.GetPath(string(keyData))
+}
+
+// CleanCache is invoked on a hard-refresh or when the manifest cache has expired. This removes the OCI image from the cached path.
+func (c *nativeOCIClient) CleanCache(revision, project string) error {
+	cachePath, err := c.getCachedPath(revision, project)
+	if err != nil {
+		return err
+	}
+	return os.RemoveAll(cachePath)
+}
+
+// DigestMetadata extracts the OCI manifest for a given revision and returns it to the caller.
+func (c *nativeOCIClient) DigestMetadata(ctx context.Context, digest, project string) (*v1.Manifest, error) {
+	path, err := c.getCachedPath(digest, project)
+	if err != nil {
+		return nil, err
+	}
+
+	repo, err := oci.NewFromTar(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	return getOCIManifest(ctx, digest, repo)
+}
+
+func (c *nativeOCIClient) ResolveRevision(ctx context.Context, revision string, noCache bool) (string, error) {
+	constraints, err := semver.NewConstraint(revision)
+	if err == nil {
+		tags, err := c.getTags(ctx, noCache)
+		if err != nil {
+			return "", fmt.Errorf("error fetching tags: %w", err)
+		}
+		version, err := tags.MaxVersion(constraints)
+		if err != nil {
+			return "", fmt.Errorf("no version for constraints: %w", err)
+		}
+		return c.resolveDigest(ctx, version.String())
+	}
+
+	return c.resolveDigest(ctx, revision)
+}
+
+func (c *nativeOCIClient) getTags(ctx context.Context, noCache bool) (*TagsList, error) {
+	indexLock.Lock(c.repoURL)
+	defer indexLock.Unlock(c.repoURL)
+
+	var data []byte
+	if !noCache && c.indexCache != nil {
+		if err := c.indexCache.GetHelmIndex(c.repoURL, &data); err != nil && !errors.Is(err, cache.ErrCacheMiss) {
+			log.Warnf("Failed to load index cache for repo: %s: %s", c.repoLock, err)
+		}
+	}
+
+	tags := &TagsList{}
+	if len(data) == 0 {
+		start := time.Now()
+		result, err := c.tagsFunc(ctx, "")
+		if err != nil {
+			return nil, fmt.Errorf("failed to get tags: %w", err)
+		}
+
+		for _, tag := range result {
+			// By convention: Change underscore (_) back to plus (+) to get valid SemVer
+			convertedTag := strings.ReplaceAll(tag, "_", "+")
+			tags.Tags = append(tags.Tags, convertedTag)
+		}
+
+		log.WithFields(
+			log.Fields{"seconds": time.Since(start).Seconds(), "repo": c.repoURL},
+		).Info("took to get tags")
+
+		if c.indexCache != nil {
+			if err := c.indexCache.SetHelmIndex(c.repoURL, data); err != nil {
```

Comment:
Same question, why HelmIndex? Should we have a new index?

Reply:
Yes, that's something which needs to be fixed. This was an artifact from taking the cache impl from Helm.